### ISSUE TITLE
构建dataset时，将instruction部分的label设置为-100可能是无效的

### ISSUE THREAD

**Issue by zhengyanzhao1997** [Facico/Chinese-Vicuna#99]
构建dataset时，将instruction部分的label设置为-100可能是无效的，原因是Trainer部分设置的transformers.DataCollatorForLanguageModeling会将label进行重制。
将
`data_collator=transformers.DataCollatorForLanguageModeling(tokenizer, mlm=False)`
替换为
`data_collator=transformers.DataCollatorForSeq2Seq(tokenizer, return_tensors="pt", padding=True)`
能解决这个问题

**Comment by Facico** [comment]
感谢，是会有这个问题，我们之后会调整

**Comment by songbaipu** [comment]
generate_and_tokenize_prompt函数中： **"attention_mask": [1] * (len(full_tokens))** 这行，看起来把padding的token也设置为1了，这个是合理的吗？

**Comment by songbaipu** [comment]
> 构建dataset时，将instruction部分的label设置为-100可能是无效的，原因是Trainer部分设置的transformers.DataCollatorForLanguageModeling会将label进行重制。 将 `data_collator=transformers.DataCollatorForLanguageModeling(tokenizer, mlm=False)` 替换为 `data_collator=transformers.DataCollatorForSeq2Seq(tokenizer, return_tensors="pt", padding=True)` 能解决这个问题

请问一下，如果不替换的话，可能导致训练会出什么错误呢？我看了很多类似的repo，貌似都和作者的写法差不多，不知道这个bug的影响是啥呢？

**Comment by zhengyanzhao1997** [comment]
> > 构建dataset时，将instruction部分的label设置为-100可能是无效的，原因是Trainer部分设置的transformers.DataCollatorForLanguageModeling会将label进行重制。 将 `data_collator=transformers.DataCollatorForLanguageModeling(tokenizer, mlm=False)` 替换为 `data_collator=transformers.DataCollatorForSeq2Seq(tokenizer, return_tensors="pt", padding=True)` 能解决这个问题
> 
> 请问一下，如果不替换的话，可能导致训练会出什么错误呢？我看了很多类似的repo，貌似都和作者的写法差不多，不知道这个bug的影响是啥呢？

没有大bug，最终性能可能会有区别，以实验结果为准哈

**Comment by zhengyanzhao1997** [comment]
> generate_and_tokenize_prompt函数中： **"attention_mask": [1] * (len(full_tokens))** 这行，看起来把padding的token也设置为1了，这个是合理的吗？

只要最后把padding id 的label 忽视掉，就没问题

**Comment by songbaipu** [comment]
> > > 构建dataset时，将instruction部分的label设置为-100可能是无效的，原因是Trainer部分设置的transformers.DataCollatorForLanguageModeling会将label进行重制。 将 `data_collator=transformers.DataCollatorForLanguageModeling(tokenizer, mlm=False)` 替换为 `data_collator=transformers.DataCollatorForSeq2Seq(tokenizer, return_tensors="pt", padding=True)` 能解决这个问题
> > 
> > 
> > 请问一下，如果不替换的话，可能导致训练会出什么错误呢？我看了很多类似的repo，貌似都和作者的写法差不多，不知道这个bug的影响是啥呢？
> 
> 没有大bug，最终性能可能会有区别，以实验结果为准哈

我大概理解了，作者原来的写法更像是语言模型自回归的训练（prmopt + input + output），因为labels并没有生效；您这边的写法就更像是QA的训练（labels部分prmopt + input的部分被mask掉了），只训练A部分对应的loss，我这样理解对吗？

**Comment by Facico** [comment]
@songbaipu 是的，我这种写法相当于是把prompt和input的生成也放进loss里面的；label起效果相当于只把output的生成放进loss里面

**Comment by Facico** [comment]
感谢，已修正。https://github.com/Facico/Chinese-Vicuna/commit/b1dd2ea9b216aa8792ccc8b6b63e3303237a186b
